# Actividad - Proyecto práctico


> La actividad se desarrollará en grupos pre-definidos de 2-3 alumnos. Se debe indicar los nombres en orden alfabético (de apellidos). Recordad que esta actividad se corresponde con un 30% de la nota final de la asignatura. Se debe entregar entregar el trabajo en la presente notebook.
*   Alumno 1: Gil Garcera, Javier
*   Alumno 2: Palomares Mateo, Abel
*   Alumno 3: Serrano López, Francisco Rubén
*   Alumno 4: Vegas Romero, David







---
## **PARTE 1** - Instalación y requisitos previos

> Las prácticas han sido preparadas para poder realizarse en el entorno de trabajo de Google Colab. Sin embargo, esta plataforma presenta ciertas incompatibilidades a la hora de visualizar la renderización en gym. Por ello, para obtener estas visualizaciones, se deberá trasladar el entorno de trabajo a local. Por ello, el presente dosier presenta instrucciones para poder trabajar en ambos entornos. Siga los siguientes pasos para un correcto funcionamiento:
1.   **LOCAL:** Preparar el enviroment, siguiendo las intrucciones detalladas en la sección *1.1.Preparar enviroment*.
2.  **AMBOS:** Modificar las variables "mount" y "drive_mount" a la carpeta de trabajo en drive en el caso de estar en Colab, y ejecturar la celda *1.2.Localizar entorno de trabajo*.
3. **COLAB:** se deberá ejecutar las celdas correspondientes al montaje de la carpeta de trabajo en Drive. Esta corresponde a la sección *1.3.Montar carpeta de datos local*.
4.  **AMBOS:** Instalar las librerías necesarias, siguiendo la sección *1.4.Instalar librerías necesarias*.


---
### 1.1. Preparar enviroment (solo local)



> Para preparar el entorno de trabajo en local, se han seguido los siguientes pasos:
1. En Windows, puede ser necesario instalar las C++ Build Tools. Para ello, siga los siguientes pasos: https://towardsdatascience.com/how-to-install-openai-gym-in-a-windows-environment-338969e24d30.
2. Instalar Anaconda
3. Siguiendo el código que se presenta comentado en la próxima celda: Crear un enviroment, cambiar la ruta de trabajo, e instalar librerías básicas.


```
conda create --name miar_rl python=3.8
conda activate miar_rl
cd "PATH_TO_FOLDER"
conda install git
pip install jupyter
```


4. Abrir la notebook con *jupyter-notebook*.



```
jupyter-notebook
```


---
### 1.2. Localizar entorno de trabajo: Google colab o local

In [57]:
# ATENCIÓN!! Modificar ruta relativa a la práctica si es distinta (drive_root)
mount='/content/gdrive'
drive_root = mount + "/My Drive/08_MIAR/actividades/proyecto practico_v4"

try:
  from google.colab import drive
  IN_COLAB=True
except:
  IN_COLAB=False

---
### 1.3. Montar carpeta de datos local (solo Colab)

In [58]:
# Switch to the directory on the Google Drive that you want to use
import os
if IN_COLAB:
  print("We're running Colab")

  if IN_COLAB:
    # Mount the Google Drive at mount
    print("Colab: mounting Google drive on ", mount)

    drive.mount(mount)

    # Create drive_root if it doesn't exist
    create_drive_root = True
    if create_drive_root:
      print("\nColab: making sure ", drive_root, " exists.")
      os.makedirs(drive_root, exist_ok=True)

    # Change to the directory
    print("\nColab: Changing directory to ", drive_root)
    %cd $drive_root
# Verify we're in the correct working directory
%pwd
print("Archivos en el directorio: ")
print(os.listdir())

We're running Colab
Colab: mounting Google drive on  /content/gdrive
Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).

Colab: making sure  /content/gdrive/My Drive/08_MIAR/actividades/proyecto practico_v4  exists.

Colab: Changing directory to  /content/gdrive/My Drive/08_MIAR/actividades/proyecto practico_v4
/content/gdrive/My Drive/08_MIAR/actividades/proyecto practico_v4
Archivos en el directorio: 
['dqn_SpaceInvaders-v0_log.json', 'dqn_SpaceInvaders-v0_weights.h5f.data-00000-of-00001', 'dqn_SpaceInvaders-v0_weights.h5f.index', 'checkpoint', 'dqn_SpaceInvaders-v0_weights_50000.h5f.data-00000-of-00001', 'dqn_SpaceInvaders-v0_weights_50000.h5f.index']


---
### 1.4. Instalar librerías necesarias

In [59]:
if IN_COLAB:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install keras-rl2==1.0.5
  %pip install tensorflow==2.12
else:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install pyglet==1.5.0
  %pip install h5py==3.1.0
  %pip install Pillow==9.5.0
  %pip install keras-rl2==1.0.5
  %pip install Keras==2.2.4
  %pip install tensorflow==2.12.1
  %pip install torch==2.0.1
  %pip install tensorflow-gpu==2.5.3
  %pip install agents==1.4.0

  Cloning https://github.com/Kojoley/atari-py.git to /tmp/pip-req-build-cxvkre37
  Running command git clone --filter=blob:none --quiet https://github.com/Kojoley/atari-py.git /tmp/pip-req-build-cxvkre37
  Resolved https://github.com/Kojoley/atari-py.git to commit 86a1e05c0a95e9e6233c3a413521fdb34ca8a089
  Preparing metadata (setup.py) ... done


---
## **PARTE 2**. Enunciado

Consideraciones a tener en cuenta:

- El entorno sobre el que trabajaremos será _SpaceInvaders-v0_ y el algoritmo que usaremos será _DQN_.

- Para nuestro ejercicio, el requisito mínimo será alcanzado cuando el agente consiga una **media de recompensa por encima de 20 puntos en modo test**. Por ello, esta media de la recompensa se calculará a partir del código de test en la última celda del notebook.

Este proyecto práctico consta de tres partes:

1.   Implementar la red neuronal que se usará en la solución
2.   Implementar las distintas piezas de la solución DQN
3.   Justificar la respuesta en relación a los resultados obtenidos

**Rúbrica**: Se valorará la originalidad en la solución aportada, así como la capacidad de discutir los resultados de forma detallada. El requisito mínimo servirá para aprobar la actividad, bajo premisa de que la discusión del resultado sera apropiada.

IMPORTANTE:

* Si no se consigue una puntuación óptima, responder sobre la mejor puntuación obtenida.
* Para entrenamientos largos, recordad que podéis usar checkpoints de vuestros modelos para retomar los entrenamientos. En este caso, recordad cambiar los parámetros adecuadamente (sobre todo los relacionados con el proceso de exploración).
* Se deberá entregar unicamente el notebook y los pesos del mejor modelo en un fichero .zip, de forma organizada.
* Cada alumno deberá de subir la solución de forma individual.

---
## **PARTE 3**. Desarrollo y preguntas

#### Importar librerías

In [68]:
import numpy as np
import gym
from PIL import Image

from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten, Conv2D, Permute
from keras.optimizers import Adam
import keras.backend as K

from rl.agents.dqn import DQNAgent
from rl.policy import LinearAnnealedPolicy, EpsGreedyQPolicy
from rl.memory import SequentialMemory
from rl.core import Processor
from rl.callbacks import FileLogger, ModelIntervalCheckpoint

#### Configuración base

In [80]:
INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4

env_name = 'SpaceInvaders-v0'
env = gym.make(env_name)

np.random.seed(123)
env.seed(123)
nb_actions = env.action_space.n

In [81]:
INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4

class CustomAtariProcessor(Processor):
    def process_observation(self, observation):
        assert observation.ndim == 3  # (210, 160, 3)
        grayscale = np.mean(observation, axis=2).astype(np.uint8) # Convierte la imagen a escala de grises haciendo la media de los canales RGB
        cropped = grayscale[30:190, 10:150]  # zona útil - Recorta la imagen para quedarse con la zona donde ocurre el juego (elimina HUD)
        img = Image.fromarray(cropped)
        img = img.resize(INPUT_SHAPE, Image.BILINEAR) # Redimensionamos la imagen a 84x84 usando interpolación bilineal
        processed = np.array(img).astype('uint8')
        return processed

    def process_state_batch(self, batch):
        return batch.astype('float32') / 255. # Normalizamos

    def process_reward(self, reward):
        return np.clip(reward, -1., 1.)

In [82]:
class RewardShapingWrapper(gym.RewardWrapper):
    def __init__(self, env):
        super().__init__(env)
        self.last_action = None

    def step(self, action):
        obs, reward, done, info = self.env.step(action)

        # Incentivamos las acciones de disparo para fomentar un comportamiento ofensivo:
        if action == 3: # Disparar Izquierda
            reward += 0.1
        elif action == 4: # Disparar Derecha
            reward += 0.2 # (mayor incentivo por el patrón observado en las primeras pruebas)
        return obs, reward, done, info

env_name = 'SpaceInvaders-v0'
env = gym.make(env_name)
env = RewardShapingWrapper(env)
np.random.seed(123)
env.seed(123)
nb_actions = env.action_space.n


**Justificación de la utilización: classe RewardShapingWrapper**

Tras visualizar los resultados en las primeras pruebas, detectamos que al agente le costaba más realizar disparos hacia la derecha, lo cual podía limitar su desempeño ofensivo. Por ello, hemos incorporado un reward shaping ligero, incentivando las acciones de disparo, especialmente hacia la derecha, para guiar al agente durante la fase de aprendizaje sin alterar el entorno original durante el test.

**IMPORTANTE**

Este ajuste tiene como consecuencia que las recompensas durante el entrenamiento estén artificialmente infladas, ya que se introducen bonificaciones externas al entorno original. Sin embargo, para obtener una evaluación objetiva, hemos realizado el test final en un entorno limpio, sin aplicar el `RewardShapingWrapper`, de forma que las recompensas finales no estén alteradas y reflejen fielmente la política aprendida por el agente.

### 1. Implementación de la red neuronal

Se ha implementado una red convolucional sencilla pero efectiva, basada en la arquitectura propuesta en el paper original de DQN para juegos de Atari. Este tipo de red permite extraer patrones espaciales relevantes de las observaciones visuales (como posiciones de enemigos o disparos), y está especialmente diseñada para procesar secuencias de imágenes apiladas (4 frames). Su estructura ligera facilita un aprendizaje eficiente sin requerir demasiados recursos computacionales, siendo adecuada para entornos visuales como SpaceInvaders-v0.

In [83]:
input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE

model = Sequential()
model.add(Permute((2, 3, 1), input_shape=input_shape))  # canales al final

model.add(Conv2D(32, (8, 8), strides=(4, 4), activation='relu'))
model.add(Conv2D(64, (4, 4), strides=(2, 2), activation='relu'))
model.add(Conv2D(64, (3, 3), strides=(1, 1), activation='relu'))
model.add(Flatten())
model.add(Dense(512, activation='relu'))
model.add(Dense(nb_actions, activation='linear'))

print(model.summary())

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 permute_6 (Permute)         (None, 84, 84, 4)         0         
                                                                 
 conv2d_18 (Conv2D)          (None, 20, 20, 32)        8224      
                                                                 
 conv2d_19 (Conv2D)          (None, 9, 9, 64)          32832     
                                                                 
 conv2d_20 (Conv2D)          (None, 7, 7, 64)          36928     
                                                                 
 flatten_6 (Flatten)         (None, 3136)              0         
                                                                 
 dense_12 (Dense)            (None, 512)               1606144   
                                                                 
 dense_13 (Dense)            (None, 6)                

### 2. Implementación de la solución DQN

In [84]:
memory = SequentialMemory(limit=1_500_000, window_length=WINDOW_LENGTH)

processor = CustomAtariProcessor()

policy = LinearAnnealedPolicy(EpsGreedyQPolicy(),
                              attr='eps',
                              value_max=1.0,
                              value_min=0.1,  # Termina con un 10% de exploración mínima
                              value_test=0.05,
                              nb_steps=int(1_500_000 * 0.8)) # ε disminuye linealmente hasta el 80% de los pasos:
                                                             # permite explorar lo suficiente antes de explotar

dqn = DQNAgent(model=model,
               nb_actions=nb_actions,
               policy=policy,
               memory=memory,
               processor=processor,
               nb_steps_warmup=50_000,
               gamma=0.99,
               target_model_update=10_000,
               train_interval=4)

dqn.compile(Adam(learning_rate=1e-4), metrics=['mae'])


### 3. Entrenamiento

Durante el entrenamiento, **se han creado dos entornos diferenciados**:

- `env_train` (entorno con reward shaping)
Se utiliza una versión modificada del entorno original (RewardShapingWrapper) que añade pequeñas recompensas extra cuando el agente dispara (más aún si lo hace hacia la derecha).
Esta modificación tiene como objetivo guiar el aprendizaje inicial incentivando un comportamiento ofensivo.

- `env_test` (entorno limpio, sin shaping)
Para garantizar una evaluación honesta, se construye un segundo entorno idéntico al original, pero sin ningún tipo de modificación.
Esto asegura que las recompensas obtenidas durante el test no estén infladas artificialmente, cumpliendo así con los requisitos del enunciado.



In [85]:
env_name = 'SpaceInvaders-v0'

# ENTRENAMIENTO (con reward shaping)
env_train = gym.make(env_name)
env_train = RewardShapingWrapper(env_train)

# TESTEO (entorno limpio, sin shaping)
env_test = gym.make(env_name)

# Semillas
np.random.seed(123)
env_train.seed(123)
env_test.seed(123)

nb_actions = env_train.action_space.n

In [86]:
weights_filename = 'dqn_{}_weights.h5f'.format(env_name)
checkpoint_weights_filename = 'dqn_' + env_name + '_weights_{step}.h5f'
log_filename = 'dqn_{}_log.json'.format(env_name)

callbacks = [
    ModelIntervalCheckpoint(checkpoint_weights_filename, interval=50000),
    FileLogger(log_filename, interval=100)
]

dqn.fit(env, callbacks=callbacks, nb_steps=1500000, log_interval=10000, visualize=False)
dqn.save_weights(weights_filename, overwrite=True)

Training for 2000000 steps ...
Interval 1 (0 steps performed)
10000/10000 [==============================] - 44s 4ms/step - reward: 0.0623
14 episodes - episode_reward: 43.657 [25.500, 71.800] - ale.lives: 2.049

Interval 2 (10000 steps performed)
10000/10000 [==============================] - 44s 4ms/step - reward: 0.0622
15 episodes - episode_reward: 41.127 [22.900, 64.100] - ale.lives: 2.135

Interval 3 (20000 steps performed)
10000/10000 [==============================] - 44s 4ms/step - reward: 0.0642
16 episodes - episode_reward: 40.938 [23.000, 81.400] - ale.lives: 2.000

Interval 4 (30000 steps performed)
10000/10000 [==============================] - 44s 4ms/step - reward: 0.0624
15 episodes - episode_reward: 40.300 [25.000, 65.600] - ale.lives: 2.152

Interval 5 (40000 steps performed)
10000/10000 [==============================] - 44s 4ms/step - reward: 0.0630
13 episodes - episode_reward: 47.623 [24.300, 93.500] - ale.lives: 2.122

Interval 6 (50000 steps performed)
10000/10

### 4. Test

In [87]:
dqn.save_weights(weights_filename, overwrite=True)

In [133]:
# Carga de pesos y evaluación
weights_filename = 'dqn_{}_weights.h5f'.format(env_name)
dqn.load_weights(weights_filename)

# Test en 10 episodios
test_history = dqn.test(env_test, nb_episodes=10, visualize=False)

# Métricas
recompensas = test_history.history['episode_reward']
pasos = test_history.history['nb_steps']

print(f"Media de la recompensa en 10 episodios: {np.mean(recompensas):.2f}")
print(f"Desviación estándar de la recompensa: {np.std(recompensas):.2f}")

print(f"Media de pasos por episodio: {np.mean(pasos):.2f}")
print(f"Desviación estándar de los pasos: {np.std(pasos):.2f}")

Testing for 10 episodes ...
Episode 1: reward: 19.000, steps: 811
Episode 2: reward: 18.000, steps: 715
Episode 3: reward: 20.000, steps: 742
Episode 4: reward: 20.000, steps: 702
Episode 5: reward: 20.000, steps: 734
Episode 6: reward: 16.000, steps: 817
Episode 7: reward: 19.000, steps: 1193
Episode 8: reward: 27.000, steps: 1515
Episode 9: reward: 21.000, steps: 1053
Episode 10: reward: 15.000, steps: 1114
Media de la recompensa en 10 episodios: 19.50
Desviación estándar de la recompensa: 3.07
Media de pasos por episodio: 939.60
Desviación estándar de los pasos: 256.58


3. Justificación de los parámetros seleccionados y de los resultados obtenidos

En este proyecto se ha entrenado un agente **DQN (Deep Q-Network)** para el entorno `SpaceInvaders-v0`, utilizando una arquitectura de red neuronal convolucional inspirada en el diseño original de DeepMind para juegos de Atari. La red incorpora tres capas `Conv2D` seguidas de `Flatten` y capas `Dense`, permitiendo una extracción jerárquica de características espaciales y la estimación de los valores Q para cada acción posible.

Durante el entrenamiento, se ha incorporado un **reward shaping ligero** que premia las acciones de disparo (izquierda y derecha), con el objetivo de acelerar la exploración de políticas ofensivas y evitar conductas pasivas en fases tempranas. Este shaping se ha desactivado completamente durante el test, siguiendo buenas prácticas para no inflar artificialmente los resultados. Para ello, se ha utilizado un entorno separado, sin modificaciones ni bonificaciones adicionales, garantizando así que las recompensas obtenidas en test reflejen exclusivamente la política aprendida por el agente.

La política de exploración empleada ha sido **ε-greedy con decaimiento lineal**, comenzando en **ε=1.0** (exploración total) y descendiendo hasta **ε=0.1** de forma progresiva. Este valor mínimo se alcanza justo al **80% del total de pasos de entrenamiento**, lo cual permite una amplia fase de exploración antes de consolidar la explotación. Esta estrategia es común en entornos complejos porque **favorece el descubrimiento de comportamientos variados al inicio**, mientras que **reserva el tramo final del entrenamiento para afinar la política con un comportamiento más determinista**.


**Resultados obtenidos (modo test)**

- **Media de recompensa en 10 episodios:** 19.50  
- **Desviación estándar de la recompensa:** 3.07  
- **Media de pasos por episodio:** 939.60  
- **Desviación estándar de pasos:** 256.58  

Estos resultados reflejan un comportamiento **estable y cercano al umbral exigido de 20 puntos**, con una desviación moderada que indica que el agente ha aprendido una política funcional y reproducible. Se observa una ligera variabilidad entre episodios, típica de entornos con cierto grado de aleatoriedad en la dinámica de los enemigos.

**Posibles líneas de mejora**

Aunque la media obtenida es sólida, se podrían considerar mejoras como:

- Aumentar ligeramente el número de pasos de entrenamiento.
- Ajustar el *learning rate* o la tasa de exploración final.
- Incorporar técnicas como **Double DQN** o **Prioritized Experience Replay**.
- Incrementar la regularización o el uso de `BatchNormalization` para mejorar la generalización.

---